Add custom CSS support

marie-bnl opened the pull request:
#### Pre-flight Checklist

- [x] I have read the [Contributing Guidelines](https://github.com/ferdium/ferdium-app/blob/develop/CONTRIBUTING.md) for this project.
- [x] I agree to follow the [Code of Conduct](https://github.com/ferdium/ferdium-app/blob/develop/CODE_OF_CONDUCT.md) that this project adheres to.
- [x] I have searched the [issue tracker](https://github.com/ferdium/ferdium-app/issues) for a feature request that matches the one I want to file, without success.

Note: I'm pretty new and don't know much about Ferdium's architecture so I tried my best, but do not hesitate to tell me if anything needs to be changed

#### Description of Change

At start, Ferdium looks for a `~/.config/Ferdium/config/custom.css` and applies it if it exists. The CSS customises Ferdium's UI, not the webview.

#### Motivation and Context

I made a custom CSS for Ferdium in order for it to better fit my desktop environment and I'd like to have a better way to apply it than just compiling the software myself.

#### Screenshots

Not very relevant, but for an example with my own theme :

![image](https://user-images.githubusercontent.com/39555268/211105487-6cccdf54-b34d-4386-aa1d-f9050ba7c1d1.png)

#### Checklist

- [x] My pull request is properly named
- [x] The changes respect the code style of the project (`pnpm prepare-code`)
- [x] `pnpm test` passes
- [x] I tested/previewed my changes locally

#### Release Notes

Added support for a custom CSS stylesheet for Ferdium's UI

SpecialAro commented:
Haven't still tested this (once I get the time I will try it - if I'm fast enough maybe I can also review and approve) but what an amazing feature to have!! 😁 Great work @victorbnl !!

I think this opens a lot of possibilities for Ferdium. I'm thinking on pre bundled themes for instance 😁

marie-bnl commented:
Well I think I didn't implement it well for themes bundling. Maybe theme bundling would better be integrated with ThemeManager or something like that. What I did is I just read the custom.css file and injected it into the app

Edit: Of course that's a beginning and if you explain me how Ferdium's styling works maybe we can make this PR morph into a full theme engine with support for both pre-bundled themes, community themes and user-created themes. That would be awesome!

vraravam reviewed (approved):
LGTM!

vraravam commented:
@SpecialAro - are you ok with merging this PR?

vraravam commented:
@victorbnl - I just noticed that some of the indentations are incorrect. Please confirm if you have run the `pnpm lint && pnpm reformat-files` scripts?

SpecialAro commented:
> @SpecialAro - are you ok with merging this PR?

I haven't test it but I can do it if it's needed 😁 nevertheless the code structure seems logic.

vraravam commented:
> I haven't test it but I can do it if it's needed 😁 nevertheless the code structure seems logic.

yes please if you can

vraravam commented:
@SpecialAro - im going to bed now - please merge this PR once you test and approve it - so that it can get into the next nightly

SpecialAro commented:
@victorbnl and @vraravam I tested it and it worked fine. But I noticed a issue: whenever you reload Ferdium (CTRL + SHIFT + R) it goes back to the normal css, therefore it stops injecting.

If this is a issue we can try to fix it before merging... otherwise we can change it later 😄

marie-bnl commented:
> @victorbnl - I just noticed that some of the indentations are incorrect. Please confirm if you have run the `pnpm lint && pnpm reformat-files` scripts?

Actually I don't think I did

Edit: I ran the command and no files changed, nor does it produce any error. So I guess my code is right already?

marie-bnl commented:
> @victorbnl and @vraravam I tested it and it worked fine. But I noticed a issue: whenever you reload Ferdium (CTRL + SHIFT + R) it goes back to the normal css, therefore it stops injecting.
> 
> If this is a issue we can try to fix it before merging... otherwise we can change it later smile

Definetely an issue that should be taken under consideration, ideally before merging in my opinion. I think I'll try to investigate it tomorrow

vraravam commented:
> Definetely an issue that should be taken under consideration, ideally before merging in my opinion. I think I'll try to investigate it tomorrow

ok - lets wait to get this fixed before we merge

marie-bnl commented:
I fixed the issue by loading the css in `webContents.on('dom-ready')` instead of `webContents.on('did-finish-load')`. As a bonus, I made it read the file in the event, which means the CSS is read again each time you reload, it goes without saying it is very useful for development.

I also changed the CSS loading method to using `webContents.insertCSS()`, which I just discovered by the way. This obviously makes the code much shorter and understandable. However, it makes it required to put `!important` at the end of each line. ~~One solution to that would be to add `cssOrigin: 'user'` to the options of insertCSS, but it breaks styling even more.~~ Nonetheless I think it is not that big of a deal as it is pretty common to have to add those `!important` when styling other applications, and it would be ready to go into production. Still, I'll try to figure out whether I can solve this issue or not.

Edit: switched back to the first method

Commit:
feat: add custom css support

vraravam commented on src/electron/customCSS.ts:
```diff
@@ -0,0 +1,14 @@
+import { readFileSync } from 'fs';
+import { pathExistsSync } from 'fs-extra';
```
I think `fs-extra` has `readFileSync` as well - you can combine the 2 import statements into a single line.

vraravam commented on src/electron/customCSS.ts:
```diff
@@ -0,0 +1,14 @@
+import { userDataPath } from '../environment-remote';
+
+export default function loadCustomCSS() {
+  const path = userDataPath('config', 'custom.css');
```
Please document this path and what this feature does in the README.md file for new users to try out if they choose to.

marie-bnl replied:
Sure

vraravam commented on src/index.ts:
```diff
@@ -263,6 +264,11 @@ const createWindow = () => {
 
+  // Inject custom CSS
+  mainWindow.webContents.on('dom-ready', () => {
+    mainWindow?.webContents.send('inject-css', loadCustomCSS());
```
Why do we need to send a new event here? Why not do the actual work in this callback?

marie-bnl replied:
Because index.ts is part of the main process and thus does not have access to the document, which only the renderer process (app.tsx) has access to, so we can't manipulate the DOM from index.ts directly

vraravam replied:
thanks - makes sense

Commit:
chore(custom css): use readFileSync from fs-extra

Commit:
docs(readme): add section about styling

Commit:
docs(readme): style

marie-bnl commented:
So on my side it seems ready to merge

vraravam commented:
@SpecialAro ?

SpecialAro commented:
Oh, I'm sorry! For some reason I'm not getting notifications from this PR 😶

Will test and merge it today!

@victorbnl thank you for your hard work on this 😁

marie-bnl commented:
No problem :D

SpecialAro commented:
I tried to test it on my windows 11 and it didn't work with the custom.css file I had before: 

![image](https://user-images.githubusercontent.com/37463445/212554029-782efe9b-4f19-4815-83a1-07213151b7eb.png)

I'll try to test even more to see what is happening. But if you wan't we can merge the PR and try to fix it later on. I can also try to test it on my macos

marie-bnl commented:
It seems like this time it was my turn not to get notifications ahah! I'm setting up a VM to see if I can reproduce it

SpecialAro commented:
> It seems like this time it was my turn not to get notifications ahah! I'm setting up a VM to see if I can reproduce it

Ahah if you want you can ping me on discord :p

I've tested on my macOS and didn't manage to get it to work neither - which is strange because it previously worked on Windows 11

SpecialAro committed:
fix: change the way we inject css

SpecialAro commented:
@victorbnl and @vraravam just pushed some changes in this PR. I've changed things a little after seeing that we already have something that allows us to inject CSS in `src/features/appearance` so I used the same file, created a function and called it when we initAppearance()

@victorbnl if you agree we can merge this PR with this solution. Otherwise, we can still try to make your code work 😄 Let me know what you think, please!

Tested with both Windows and macOS - working fine 🤓 

Thank you both!

marie-bnl commented:
Great! It works on my Linux

vraravam committed:
Minor refactoring

vraravam commented on src/features/appearance/index.ts:
```diff
@@ -20,6 +21,15 @@ function createStyleElement() {
+function injectUserCustomCSS() {
+  const contentToInject = loadCustomCSS();
+  if (contentToInject !== '') {
+    const styleElement = document.createElement('style');
```
Why can't we use the `setAppearance` function to override with the user's custom style?

SpecialAro replied:
We sure can, but I think we need to change setAppearance a little so it handles a different id (I have the opinion this style should be under a different id than `custom-appearance-style`, so it is clear that comes from a user file)

vraravam replied:
ok - that makes sense. Yes, we can refactor to have 2 different css-ids.

But, the original one itself proclaims as "custom" - what is it a "custom" style of? Or should that one be called "main" or "global"  or "default" or something similar, and this new one called "custom"?

vraravam commented on src/features/appearance/index.ts:
```diff
@@ -20,6 +21,15 @@ function createStyleElement() {
 }
 
+function injectUserCustomCSS() {
+  const contentToInject = loadCustomCSS();
```
Is it better/more maintainable to inline `loadCustomCSS` in this file/line rather than keeping it in a separate file?

SpecialAro replied:
I agree, moving it now

SpecialAro committed:
refactor: move customCSS logic; add function to generateStyle

vraravam commented:
@all-contributors please add @victorbnl for code
